### PR TITLE
doc(storage): clarify `ObjectWriteStream` behavior

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11015)

### CODE REVIEW

File: google/cloud/storage/object_write_stream.h [hunk 1]
```diff
@@ -33,63 +33,91 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
...
- * * This API is designed for applications that need to stream the object
+ * - This API is designed for applications that need to stream the object
  *   payload. If you have the payload as one large buffer consider using
- *   `storage::Client::InsertObject()`, it is simpler and faster in most cases.
- * * This API can be used to perform unformatted I/O, as well as formatted I/O
- *   using the familiar `operator<<` APIs. Note that formatted I/O typically
- *   implies some form of buffering and data copying. For best performance,
- *   consider using the [.write()][cpp-reference-write] member function.
- * * GCS expects to receive data in multiples of the *upload quantum* (256KiB).
+ *   `Client::InsertObject()`, it is simpler and faster in most cases.
```
Comment (devbww) [inline google/cloud/storage/object_write_stream.h:38]: s/,/;/

File: google/cloud/storage/object_write_stream.h [hunk 2]
```diff
@@ -33,63 +33,91 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
...
  *   Sending a buffer that is not a multiple of this quantum terminates the
- *   upload. This constraints the implementation of buffered and unbuffered I/O
- *   as described below.
+ *   upload.
+ *   - Consequently, this class must maintain an internal buffer before sending
+ *     the data to the service.
+ *   - Understanding how this buffer is used is important to get the best
+ *     possible performance.
+ *   - When using unformatted I/O, try to size your data in multiples of the
+ *     upload quantum, this often results in better performance.
```
Comment (devbww) [inline google/cloud/storage/object_write_stream.h:53]: s/, / as /

File: google/cloud/storage/object_write_stream.h [hunk 3]
```diff
@@ -33,63 +33,91 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
...
+ * @ref Client object is created. The current default value is 8 MiB, but this
+ * default value can change. If the size of this buffer is important for your
+ * application please set the value explicitly. You can also provide an override
+ * when calling `Client::WriteObject()`. Note that this setting is expressed in
+ * bytes, but it is always rounded (up) to an upload quantum.
+ *
+ * #### Unformatted I/O
+ *
+ * On a `.write()` call this class attempts to send the data immediately,
+ * that is, without copying it to the internal buffer. If any previously
```
Comment (devbww) [inline google/cloud/storage/object_write_stream.h:66]: s/, that is/. That is/

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/object_write_stream.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/object_write_stream.h]: Fixed

File: google/cloud/storage/object_write_stream.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/object_write_stream.h]: Fixed

File: google/cloud/storage/object_write_stream.h [hunk 4]
```diff
@@ -33,63 +33,91 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
...
- * data provided in the `.write()` call are larger than an upload quantum the
- * class sends data immediately. Any data in excess of a multiple of the upload
- * quantum are buffered for the next upload.
+ * The maximum size of this internal buffer is configured using
+ * `UploadBufferSizeOption`. As with all options, this can be set when the
+ * @ref Client object is created. The current default value is 8 MiB, but this
+ * default value can change. If the size of this buffer is important for your
+ * application please set the value explicitly. You can also provide an override
+ * when calling `Client::WriteObject()`. Note that this setting is expressed in
+ * bytes, but it is always rounded (up) to an upload quantum.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_write_stream.h]: Yup

File: google/cloud/storage/object_write_stream.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/object_write_stream.h]: Fixed.  Thanks for the thorough reviews (as usual).

**Pull request merged**